### DISCUSSION TITLE
How to create type generics for coin?

### DISCUSSION THREAD

**Discussion by Metal-joker** [aptos-labs/aptos-developer-discussions#146] [Questions] [answered]
currently the coin api is like this:
```
    public fun initialize<CoinType>(
        account: &signer,
        name: string::String,
        symbol: string::String,
        decimals: u8,
        monitor_supply: bool,
    ): (BurnCapability<CoinType>, FreezeCapability<CoinType>, MintCapability<CoinType>)
```
when we need to create a coin, we need to define a `CoinType` and call the initialize,
```
struct MyCoinType {}
coin::initialize<MyCoinType>();
```
what if we want to create a function that can create multiple coins?
the dex LP coins solution is to add type generics to the `CoinType`
```
struct LP<X, Y> {}
coin::initialize<LP<X, Y>>();
```
But how to get the `X` and `Y`? I'm confused.

**Comment by gregnazario** [comment, marked as answer]
Most of the DEXes will create an LP that is `LP<X,Y>`.  To create this, you need to deploy a struct in a contract like so:
```move
module 0x42::example {
  struct LP<phantom X, phantom Y> {}
}
```

Then you can use it in code and pass in the type arguments X and Y into the entry function type arguments:

```move
module 0x42::example {
    entry fun create_coin<X, Y>() {
        coin::initialize<LP<X, Y>>();
    }
}
```
